examples/tests/sdk_tests/std/array.test.w

```diff
@@ -399,3 +399,65 @@ test "removeFirst()" {
   assert(r2 == false);
   assert(mutArr.length == 3);
 }
+
+//-----------------------------------------------------------------------------
+// sort() for a mutable array
+
+test "sort()" {
+    let mutArr = MutArray<num>[3, 1, 2];
+    mutArr.sort();
+
+    assert(mutArr == MutArray<num>[1, 2, 3]);
+}
+
+//-----------------------------------------------------------------------------
+// sorted() for a non-mutable array
+
+test "sorted()" {
+    let arr = Array<num>[3, 1, 2];
+    let sortedArr = arr.sorted();
+
+    assert(sortedArr == Array<num>[1, 2, 3]);
+    assert(arr == Array<num>[3, 1, 2]);
+}
+
+//-----------------------------------------------------------------------------
+// sorted() for a mutable array with a comparator
+
+test "sortedMutWithComparator()" {
+  let mutArr = MutArray<num>[3, 2, 1];
+  mutArr.sort((a, b) => {
+    if (a < b) {
+        return -1;
+    }
+
+    if (a > b) {
+        return 1;
+    }
+
+    return 0;
+  });
```

Comment:
Sounds good; I'll go ahead with implementing the `sorted` method in `MutArray` as well. In addition, I was trying to use an structure to sort using a field and I got an error of **property not found**, investigating a bit on the project I saw it was a bug a few months ago for another particular use case so this might be a bug on the compiler side

First of all, the compiler throws an error regarding `Unable to infer type` when using an `struct` and on the other hand, if you try to access the property of the object, it throws: `Property not found`

E.g:

```
  let person = Person {
    age: 25
  };

  let person2 = Person {
    age: 30
  };

  let person3 = Person {
    age: 35
  };

  let mutArr = MutArray<Person>[person3, person, person2];

  // implement a sort by age in desc order
  let sortedArr = mutArr.sorted((a, b) => {
    if a.age > b.age {
      return -1;
    }

    if a.age < b.age {
      return 1;
    }

    return 0;
  });
```

Errors generated after running the piece of code paste above:
<img width="566" alt="image" src="https://github.com/winglang/wing/assets/13340320/64e90b97-d796-4ee7-a300-c0780f3f407c">

Reply:
Does it work if you annotate the types?

Reply:
Yeah, with explicit type annotation inside the callback works as a charm although I thought I could be implicit about the type given I'm using generic at `MutArray` level

Reply:
This looks like an inference bug. Replicated with this:

```wing
let sorted = (fn: ((num): void)?) => {};

sorted((a) => {});
//      ^ can't infer this
```

Created this here and I'll look into this now https://github.com/winglang/wing/issues/5680

Reply:
Thanks, Mark; additionally, a callback with annotations does not sort in the end, so my guess here is that the type checker is passing through, although failing silently because the array itself does not get sorted after passing the following callback

```
test "sortedMutWithComparatorAndStructs()" {
  let mutArr = MutArray<Person>[Person{age: 30}, Person{age: 40}, Person{age: 20}];

  // implement a reverse sort (instead of the default ascending sort)
  let sortedArr = mutArr.sorted((internalA, internalB) => {
    let a : Person = internalA;
    let b : Person = internalB;

    if a.age > b.age {
      return -1;
    }

    if a.age < b.age {
      return 1;
    }

    return 0;
  });

  log("{Json.stringify(sortedArr)}");
}
```

Reply:
Are you saying that this snippet doesn't work when annotation? `log("{Json.stringify(sortedArr)}");` shows an unsorted list? My **PR** to fix the inference is merged but I'm not sure this is a type issue.

By the way, I really appreciate your patience with all the changes needed to get all this to work. Implementing stuff like this properly requires stretching the language in important places, so we're glad you're up for it.

Reply:
Exactly! No matter what callback I pass through, the array remains unaffected. Apparently, (debugging the .wing artifacts), I can replicate the non-sorting issue (in Javascript), so something weird might be going on in the translation step from the SDK to the Javascript closure.

This is the code generated using `struct` as types (it works for `num`, `string`, etc)
```
module.exports = function({  }) {
  class $Closure24 {
    constructor({  }) {
      const $obj = (...args) => this.handle(...args);
      Object.setPrototypeOf($obj, this);
      return $obj;
    }
    async handle() {
      const mutArr = [({"age": 30}), ({"age": 40}), ({"age": 20})];
      const sortedArr = ((arr, args) => { return args && args[0] ? [...arr].sort(args[0]) : [...arr].sort() })(mutArr, (async (a, b) => {
        if ((a.age > b.age)) {
          return (-1);
        }
        if ((a.age < b.age)) {
          return 1;
        }
        return 0;
      }));
      $helpers.assert($helpers.eq(mutArr, [({"age": 30}), ({"age": 40}), ({"age": 20})]), "mutArr == MutArray<Person>[Person{age: 30}, Person{age: 40}, Person{age: 20}]");
      $helpers.assert($helpers.eq(sortedArr, [({"age": 20}), ({"age": 30}), ({"age": 40})]), "sortedArr == Array<Person>[Person{age: 20}, Person{age: 30}, Person{age: 40}]");
    }
  }
  return $Closure24;
}
```

I will continue to investigate. As for patience, I'm here to learn how salsa is made, and all this back and forth makes me incredibly happy because I get to see with my own eyes how a language like this is made. It's a really fun experience for me and my ideal is to one day be able to contribute to these things without screwing up everyone's schedule 😆

Reply:
@Warkanlock keep on the good work!

Reply:
Found the issue; it's the way "args" it's being read inside the `macro`

I'll investigate further

Reply:
The issue is here, where the callback is async

```js
      const sortedArr = ((arr, args) => { return args && args[0] ? [...arr].sort(args[0]) : [...arr].sort() })(mutArr, (async (a, b) => {
        if ((a.age > b.age)) {
          return (-1);
        }
        if ((a.age < b.age)) {
          return 1;
        }
        return 0;
      }));
```

It's async because all inflight closures are async and `.sort()` can't handle an async callback. We can't just strip the async off (technically the closure could be `await`ing even though that would be a bad thing to do in a comparator), so we need a way to sort with an async comparator.